cider/tests/utils/ArrowArrayBuilder.h

```diff
@@ -0,0 +1,244 @@
+    null_vecs_.push_back(null_data);
+    if (!is_row_num_set_ ||  // have not set row num, use this col_data's row num
+        row_num_ == 0) {     // previous columns are all empty
+      is_row_num_set_ = true;
+      row_num_ = col_data.size();
+    }
+    int8_t* buf = nullptr;
+    if (col_data.empty()) {
+      // append an empty buffer
+    } else {
+      // check row num
+      if (row_num_ != col_data.size()) {
+        CIDER_THROW(CiderCompileException, "Row num is not equal to previous columns!");
+      }
+      CHECK_EQ(row_num_, col_data.size());
+      // check null data num
+      if (!null_data.empty()) {
+        CHECK_EQ(row_num_, null_data.size());
+      }
+      buf = allocator_->allocate(sizeof(int64_t) * row_num_);
+      int64_t* dump = reinterpret_cast<int64_t*>(buf);
+      for (auto i = 0; i < row_num_; i++) {
+        dump[i] = col_data[i].getInt64Val();
+      }
+    }
+    table_ptr_.push_back(buf);
+    return *this;
+  }
+
+  template <typename T,
+            std::enable_if_t<std::is_same<T, CiderByteArray>::value, bool> = true>
+  ArrowArrayBuilder& addColumn(const std::string& col_name,
+                               const ::substrait::Type& col_type,
+                               const std::vector<CiderByteArray>& col_data,
+                               const std::vector<bool>& null_data = {}) {
+    col_names_.push_back(col_name);
+    col_types_.push_back(col_type);
+    null_vecs_.push_back(null_data);
+    if (!is_row_num_set_ ||  // have not set row num, use this col_data's row num
+        row_num_ == 0) {     // previous columns are all empty
+      is_row_num_set_ = true;
+      row_num_ = col_data.size();
+    }
+    int8_t* buf = nullptr;
+    if (!col_data.empty()) {
+      // check row num
+      if (row_num_ != col_data.size()) {
+        CIDER_THROW(CiderCompileException, "Row num is not equal to previous columns!");
+      }
+      CHECK_EQ(row_num_, col_data.size());
+      // check null data num
+      if (!null_data.empty()) {
+        CHECK_EQ(row_num_, null_data.size());
+      }
+      int len = sizeof(CiderByteArray) * row_num_;
+      buf = allocator_->allocate(len);
+      std::memcpy(buf, col_data.data(), len);
+    }
+    table_ptr_.push_back(buf);
+    return *this;
+  }
+
+  CiderBatch build() {
+    if (!is_row_num_set_) {
+      CIDER_THROW(CiderCompileException, "Invalid build!");
+    }
+    std::shared_ptr<CiderTableSchema> schema =
+        std::make_shared<CiderTableSchema>(col_names_, col_types_, table_name_);
+    CiderBatch batch = CiderBatch(row_num_, schema);
+    for (int i = 0; i < table_ptr_.size(); i++) {
+      if (table_ptr_[i]) {
+        std::memcpy(const_cast<int8_t*>(batch.column(i)),
+                    table_ptr_[i],
+                    batch.column_type_size(i) * row_num_);
+        std::free(const_cast<int8_t*>(table_ptr_[i]));
+      }
+    }
+    batch.set_null_vecs(null_vecs_);
+
+    convertToArrowRepresentation();
```

Comment:
We should not use this `convertToArrowRepresentation` but directly create an arrowarray.